### DISCUSSION TITLE
JS-Native Interaction Support

### DISCUSSION THREAD

**Discussion by MillyRen** [KevinnZou/compose-webview-multiplatform#60] [General]
Hey KevinnZou team, it's me again,

🌟 Quick question about your awesome library. Does it support JS calling to native side functions and vice versa, like the JS bridge?

Thanks a bunch! 🚀
Cheers

**Comment by KevinnZou** [comment]
@MillyRen Hi! Currently, we only support native calls to JS by `navigator.evaluateJavascript`. However, we have been making efforts to add support for JsBridge. I have just completed the basic functionality of JsBridge, and you can test it on this [branch](https://github.com/KevinnZou/compose-webview-multiplatform/tree/feature/jsbridge_support). I would appreciate it if you could provide any suggestions for my implementation. I will begin testing next week and, if successful, aim to release it before the month ends. You can subscribe to this [issue](https://github.com/KevinnZou/compose-webview-multiplatform/issues/31) to get the latest updates. Thank you!

**Reply by MillyRen** [reply]
Sure I will! 
and you're truly a genius!!🥹🥹🥹

**Comment by nickfaces** [comment]
Is it possible to build a test assembly using jitpack.io ? Or another simple method to add in project?

**Reply by KevinnZou** [reply]
Sure, I will release a snapshot version tomorrow.

**Reply by nickfaces** [reply]
Thank you. Unfortunately, my attempts to build the library using jitpack were unsuccessful (

**Reply by KevinnZou** [reply]
@nickfaces Hi, I have released version 1.8.0-SNAPSHOT for testing this new feature.
To import it, you have to add this repo: maven("https://s01.oss.sonatype.org/content/repositories/snapshots")

For instructions on how to use it, please refer to this [link](https://github.com/KevinnZou/compose-webview-multiplatform/tree/feature/jsbridge_support?tab=readme-ov-file#communication-between-webview-and-native).

Please let me know if you encounter any problems or if you would like any new features. Thanks!

**Reply by nickfaces** [reply]
It's a great job! It`s works fine on android and ios.  A small question. We receive a message in 'override fun handle' in a class based on 'IJsMessageHandler'. 
This is a non-composable function. The only way to return the message to the composable function, in which the webview is located, is to use something like LiveData?

**Reply by nickfaces** [reply]
For example. From webview I receive message with "edit_mode" on/off . Depending on the state of the edit_mode, I need to change the url in the webview.

**Reply by KevinnZou** [reply]
@nickfaces I got your point and agree that this is a problem that needs to be addressed. I am considering adding the webviewNavigator in the `handle` method so that you can use it to reload the webview. However, we still need to figure out how to support other UI update needs.

One possible solution is to maintain a flow inside the WebViewState. Then, `JSBridge` not only dispatches the JS call to `IJsMessageHandler`, but also to WebViewState. Then, you can collect that flow to receive the JS call and update the UI.

Do you have any other suggestions? Thank you for your feedback again!

**Reply by nickfaces** [reply]
@KevinnZou Thanks! There is still such a moment. When you click on a button, I need to change, for example, not the url, but switch to another screen (using the same voyager, for example). If I understand correctly, at the moment I cannot perform such an operation.  This is even more necessary than just changing the url in webview . 
And thank you again for the speed of response and the speed of implementing the necessary features, and even just for the good communication and help! You are really my man of year)

**Reply by KevinnZou** [reply]
@nickfaces I understand what you mean, and this is indeed a problem. This may only be resolved by creating a flow in WebViewState, as mentioned above. Since each developer uses a different navigation library, we cannot provide built-in support for navigation capabilities within the framework. Therefore, developers can only receive JS calls in the UI and handle them accordingly through flows. 

However, I haven't decided yet whether to incorporate this flow in WebViewState, because developers can actually implement a simple EventBus using shared flow to handle events. Additionally, I came across a KMP [EventBus](https://github.com/DevSrSouza/EventKt) library that should be able to help you send events in the handler and receive notifications in the UI for you to handle. Interestingly, the author of this library is also the author of Voyager, so it might be worth giving it a try. This way, you won't have to rely on the handler to provide navigation capabilities. What do you think?

Thank you for your helpful suggestions once again. You really helped me a lot and I truly appreciate it!

**Reply by nickfaces** [reply]
@KevinnZou  I'll try to figure something out with this... Thanks.
I tried to implement this through SharedFlow, but I stumbled on the fact that I just used LifecyclerOwner on android (not a multiplatform), in this case I am a little confused. I'm sorry if I'm asking a stupid question, but Kotlin is not my main language and I have to deal with it.
Therefore, I would be very grateful for a clear example and for any hint at all.

p.s. EventKt library is not in repository

**Reply by KevinnZou** [reply]
@nickfaces To easily post and receive events in your project, simply import the EventKt library. If Kotlin is not your primary language, it may be challenging for you to implement this functionality on your own in KMM. If you are interested in it, you can refer to this [sample](https://gist.github.com/Tetraquark/b5f2f5038aed62c9bd6cab4bb05c0a6b).

**Reply by nickfaces** [reply]
@KevinnZou Unfortunately, I can't just import EventKt because the developer's local repository doesn't work. Thanks again, I'll try to do something.

**Reply by KevinnZou** [reply]
@nickfaces I just realized that it is indeed not possible to download. I saw that you have already raised the issue with him. I'm sorry for the trouble caused and let's wait for the author to upload it to Maven Central first. I will also try to implement a simple event bus myself and push it to this branch.

**Reply by KevinnZou** [reply]
@nickfaces I have implemented a simple EventBus and pushed it to [this branch](https://github.com/KevinnZou/compose-webview-multiplatform/commit/aca61b1db7d3f0db08b639cfbd8b42f7ac1d6806)

**Reply by nickfaces** [reply]
@KevinnZou Great! Thanks a lot!

**Reply by MillyRen** [reply]
Hi Kevinn Zou @KevinnZou ,

The scenario is as follows: when receiving a JS bridge named 'openurl' in a WebView, I need to launch a new WebView. 
Can I add a method to launch a new WebView in the FlowEventBus callback? 
I tried it in the 1.8.0 sample, but it seems that I cannot call a @Composable function in the event callback because it is a suspend function.

**Reply by KevinnZou** [reply]
@MillyRen Could you attach the code on how you launch a new WebView?

**Reply by MillyRen** [reply]
Hi Kevinn Zou @KevinnZou , thank you very much for your response. Here's my code below. I apologize if I'm causing any inconvenience; I'm fairly new to Kotlin and KMP, so I genuinely apologize if this is a bother 😂

<img width="595" alt="image" src="https://github.com/KevinnZou/compose-webview-multiplatform/assets/152940344/cf91b683-88c9-4cc6-88fa-82a5ba9e800b">

or 

<img width="782" alt="image" src="https://github.com/KevinnZou/compose-webview-multiplatform/assets/152940344/d9fa5b78-cc7a-4827-bcec-d9b7fa263992">

**Reply by KevinnZou** [reply]
@MillyRen Hello, I think I understand your question now. As a responsive UI framework, Compose requires all UI changes to be triggered by changing the state, which automatically updates the UI. Therefore, you cannot directly launch a new WebView by calling a Composable method. I'm not sure if your requirement is to load a new URL or open a new page. If you only need to load a new URL, you can do so using `navigator.loadUrl`. If you need to open a new page, you will need to use a navigation framework such as Voyager.

I would recommend starting with the [introductory](https://developer.android.com/courses/pathways/jetpack-compose-for-android-developers-3) documentation of Jetpack Compose, as it provides many helpful tutorials to help you understand reactive UI.

**Reply by MillyRen** [reply]
you are truly superman🥹
